stubs/oauthlib/oauthlib/common.pyi

```diff
@@ -1,64 +1,81 @@
-from _typeshed import Incomplete
-from typing import Any
+import re
+from collections.abc import Iterable, Mapping, Sequence
+from logging import Logger
+from typing import Any, Final, Literal, TypeVar, overload
+from typing_extensions import TypeAlias
 
-UNICODE_ASCII_CHARACTER_SET: str
-CLIENT_ID_CHARACTER_SET: str
-SANITIZE_PATTERN: Any
-INVALID_HEX_PATTERN: Any
-always_safe: str
-log: Any
+_T = TypeVar("_T")
+_K = TypeVar("_K")
+_V = TypeVar("_V")
 
-def quote(s, safe: bytes = b"/"): ...
-def unquote(s): ...
-def urlencode(params): ...
-def encode_params_utf8(params): ...
-def decode_params_utf8(params): ...
+_HTTPMethod: TypeAlias = Literal["CONNECT", "DELETE", "GET", "HEAD", "OPTIONS", "PATCH", "POST", "PUT", "TRACE"]
 
-urlencoded: Any
+UNICODE_ASCII_CHARACTER_SET: Final[str]
+CLIENT_ID_CHARACTER_SET: Final[str]
+SANITIZE_PATTERN: Final[re.Pattern[str]]
+INVALID_HEX_PATTERN: Final[re.Pattern[str]]
+always_safe: Final[str]
+log: Logger
 
-def urldecode(query): ...
-def extract_params(raw): ...
-def generate_nonce(): ...
-def generate_timestamp(): ...
-def generate_token(length: int = 30, chars=...): ...
-def generate_signed_token(private_pem, request): ...
+def quote(s: str | bytes, safe: bytes = b"/") -> str: ...
+def unquote(s: str | bytes) -> str: ...
+def urlencode(params: list[tuple[str | bytes, str | bytes]]) -> str: ...
+def encode_params_utf8(params: list[tuple[str | bytes, str | bytes]]) -> list[tuple[bytes, bytes]]: ...
+def decode_params_utf8(params: list[tuple[str | bytes, str | bytes]]) -> list[tuple[str, str]]: ...
+
+urlencoded: Final[set[str]]
+
+def urldecode(query: str | bytes) -> list[tuple[str, str]]: ...
+def extract_params(raw: str | bytes | dict[str, str] | Iterable[tuple[str, str]]) -> list[tuple[str, str]] | None: ...
+def generate_nonce() -> str: ...
+def generate_timestamp() -> str: ...
+def generate_token(length: int = 30, chars: Sequence[str] = ...) -> str: ...
+def generate_signed_token(private_pem: str, request) -> str: ...
 def verify_signed_token(public_pem, token): ...
-def generate_client_id(length: int = 30, chars=...): ...
-def add_params_to_qs(query, params): ...
-def add_params_to_uri(uri, params, fragment: bool = False): ...
-def safe_string_equals(a, b): ...
-def to_unicode(data, encoding: str = "UTF-8"): ...
+def generate_client_id(length: int = 30, chars: Sequence[str] = ...) -> str: ...
+def add_params_to_qs(query: str, params: dict[str, str] | Iterable[tuple[str, str]]) -> str: ...
+def add_params_to_uri(uri: str, params: dict[str, str] | Iterable[tuple[str, str]], fragment: bool = False) -> str: ...
+def safe_string_equals(a: str, b: str) -> bool: ...
+@overload
+def to_unicode(data: str | bytes, encoding: str = "UTF-8") -> str: ...
+@overload
+def to_unicode(data: Mapping[_K, _V], encoding: str = "UTF-8") -> dict[str, _V]: ...
```

Comment:
We *could* probably be more inclusive than just `Mapping` here, but I'm with it for now for simplicity's sake. That said, since `_K` is only used once, the type var doesn't make much sense. To prevent variance issues, I'd use the following:

```suggestion
def to_unicode(data: Mapping[str, _V] | Mapping[bytes, _V], encoding: str = "UTF-8") -> dict[str, _V]: ...
```

(Although technically, mappings with mixed key types are allowed as well, I think it's a use case we can safely ignore.)